File: src/002/p002.cpp
#include <iostream>
#include <stdlib.h>
#include <math.h>

using namespace std;

#define LIMIT 4000000
typedef long double LD;
typedef unsigned long long ULL;

LD phi = (sqrt(5) + 1) / 2;
LD phi_inverse = phi - 1;

ULL fib(int i) {
    return (ULL) ((pow(phi, i) - pow(-phi_inverse, i)) / sqrt(5));
}

ULL solve() {
    ULL result = 0;
    ULL fn = fib(2);
    for (int i = 3; fn < LIMIT; i++) {
        if (fn % 2 == 0) {
            result += fn;
        }
        fn = fib(i);
    }
    return result;
}

int main() {
    ULL result = solve();
    cout << result << endl;
}

File: src/002/p002.scala
import scala.math.pow
import scala.math.sqrt

object Main {
  def fib(n: Int): Int = {
    ((pow((1 + sqrt(5)), n) - pow((1 - sqrt(5)), n)) / (pow(2, n) * sqrt(5))).toInt
  }

  def solve(): Int = {
    var n = 0
    var fibValue = 0
    var fibSum = 0

    while (fibValue < 4e6) {
      if (fibValue % 2 == 0) fibSum += fibValue
      n += 1
      fibValue = fib(n)
    }

    fibSum
  }

  def main(args: Array[String]): Unit = {
    val result = solve()
    println(result)
  }
}
